mesonbuild/modules/qt.py

```diff
@@ -122,14 +122,16 @@ def preprocess(self, state, args, kwargs):
             sources.append(ui_output)
         inc = get_include_args(include_dirs=include_directories)
         if len(moc_headers) > 0:
+            arguments = moc_extra_arguments + inc + ['@INPUT@', '-o', '@OUTPUT@']
             moc_kwargs = {'output': 'moc_@BASENAME@.cpp',
-                          'arguments': inc + ['@INPUT@', '-o', '@OUTPUT@']}
+                          'arguments': arguments}
             moc_gen = build.Generator([self.moc], moc_kwargs)
             moc_output = moc_gen.process_files('Qt{} moc header'.format(self.qt_version), moc_headers, state)
             sources.append(moc_output)
         if len(moc_sources) > 0:
+            arguments = [moc_extra_arguments, '@INPUT@', '-o', '@OUTPUT@']
```

Comment:
This is why your code fails now, I understand now your previous comment about string argument stuff. 
moc_extra_arguments is extracted from kwargs with extract_as_list function which ensure that it is either an empty list or a list with the expected values. 
So to build args you must do 'moc_extra_arguments  + ['@INPUT@', '-o', '@OUTPUT@']' to get a flat list.
Sorry I didn't see that before.